File: src/SkyBlock/enchants/armor/Virtuous.php
<?php


namespace SkyBlock\enchants\armor;


use pocketmine\player\Player;

class Virtuous extends BaseArmorEnchant {

    static int $id = 133;

    /**
     * @param Player $attacker
     * @param Player $victim
     * @param int    $enchantmentLevel
     */
    public function onActivation(Player $attacker, Player $victim, int $enchantmentLevel) : void {
        $i = 0;
        $this->sendActivation($victim, "§bVirtuous §aActivated!");
        $this->sendActivation($attacker, "§cStruck by §bVirtuous §cEnchant!");
        $limit = ceil($this->getLevel($enchantmentLevel) / 2);
        foreach ($victim->getEffects()->all() as $effect) {
            if ($i < $limit) {
                if ($effect->getType()->isBad()) {
                    $victim->getEffects()->remove($effect->getType());
                    $i++;
                }
            }
        }
    }

}

File: SkyblockCore/src/main/java/dev/skyblock/enchants/armor/Virtuous.java
package dev.skyblock.enchants.armor;

import dev.skyblock.enchants.BaseArmorEnchant;
import org.bukkit.entity.Player;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.potion.PotionEffect;
import java.util.concurrent.ThreadLocalRandom;

public class Virtuous extends BaseArmorEnchant {
    public static final int ID = 133;

    public Virtuous() {
        super(ID, "Virtuous", 15);
    }

    @Override
    public boolean isApplicableTo(Player player, int level) {
        return ThreadLocalRandom.current().nextInt(1, 26) == 1;
    }

    @Override
    public void onActivation(Player victim, EntityDamageByEntityEvent ev, int level) {
        victim.sendMessage("§bVirtuous §aActivated!");
        int i = 0;
        int limit = (int) Math.ceil(Math.ceil(level / 2.0) / 2.0);
        for (PotionEffect effect : victim.getActivePotionEffects()) {
            if (i >= limit) break;
            // Best effort to determine if bad, many are obviously bad
            if (isBadEffect(effect)) {
                victim.removePotionEffect(effect.getType());
                i++;
            }
        }
    }

    private boolean isBadEffect(PotionEffect effect) {
        String name = effect.getType().getName();
        return name.contains("POISON") || name.contains("WITHER") || name.contains("SLOW") || name.contains("BLIND") || name.contains("CONFUSION") || name.contains("WEAKNESS") || name.contains("HUNGER");
    }
}
